File: android/app/src/main/java/com/pigeon/mobile/auth/AuthViewModel.kt
package com.pigeon.mobile.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import uniffi.pigeon_mobile_core.CoreException
import uniffi.pigeon_mobile_core.PigeonClient
import uniffi.pigeon_mobile_core.Session
import uniffi.pigeon_mobile_core.login as coreLogin
import uniffi.pigeon_mobile_core.register as coreRegister
import uniffi.pigeon_mobile_core.restoreSession as coreRestoreSession

/** UI state for the auth flow. */
sealed interface AuthState {
    /** Checking for a persisted session on launch. */
    data object Restoring : AuthState

    /** Signed out — showing the form. [error] holds the last failure, if any. */
    data class SignedOut(val error: String? = null) : AuthState

    /** A register/login is in flight. */
    data object Submitting : AuthState

    /**
     * Signed in with this session identity. [client] is the opaque core handle
     * the rooms/sync flows hang off (the token stays inside it — Gotcha #1).
     * [signingOut] is true while a logout is in flight; [error] carries a logout
     * failure that left the session intact (so the user can retry rather than see
     * a signed-out UI over a live session).
     */
    data class SignedIn(
        val session: Session,
        val client: PigeonClient,
        val signingOut: Boolean = false,
        val error: String? = null,
    ) : AuthState
}

/**
 * The thin view-model over the core's session API. It owns no protocol or crypto
 * logic — it calls the core's suspend functions, holds the resulting client
 * handle, and exposes UI state. (CLAUDE.md: native = UI + a thin view-model.)
 */
class AuthViewModel : ViewModel() {

    private val _state = MutableStateFlow<AuthState>(AuthState.Restoring)
    val state: StateFlow<AuthState> = _state.asStateFlow()

    // The logged-in client handle. The token stays inside it (in the core) — this
    // is an opaque handle, kept for the flows that hang off it in later phases
    // (sync, logout); never unwrapped into app-level secret state.
    private var client: PigeonClient? = null

    init {
        // Restore a persisted session on launch.
        viewModelScope.launch {
            _state.value = try {
                coreRestoreSession()?.let { restored ->
                    client = restored
                    AuthState.SignedIn(restored.session(), restored)
                } ?: AuthState.SignedOut()
            } catch (e: CoreException) {
                // A restore fault (e.g. storage) must not wedge launch — fall
                // back to the sign-in form and surface the reason.
                AuthState.SignedOut(authErrorMessage(e))
            }
        }
    }

    fun login(server: String, username: String, password: String) =
        submit { coreLogin(server.trim(), username.trim(), password) }

    fun register(server: String, username: String, password: String) =
        submit { coreRegister(server.trim(), username.trim(), password) }

    /**
     * Sign out: revoke the token server-side and clear the persisted session
     * (both in the core). On success the handle is dropped and we return to the
     * form. The core clears local state even if the server revoke fails; the only
     * error it surfaces is a keystore fault — in that case the session is still
     * live, so we stay signed in and show the reason for a retry.
     */
    fun logout() {
        val current = _state.value
        if (current !is AuthState.SignedIn || current.signingOut) return
        val c = client ?: return
        _state.value = current.copy(signingOut = true, error = null)
        viewModelScope.launch {
            _state.value = try {
                c.logout()
                client = null
                AuthState.SignedOut()
            } catch (e: CoreException) {
                current.copy(signingOut = false, error = authErrorMessage(e))
            }
        }
    }

    private fun submit(call: suspend () -> PigeonClient) {
        if (_state.value == AuthState.Submitting) return
        _state.value = AuthState.Submitting
        viewModelScope.launch {
            _state.value = try {
                val c = call()
                client = c
                AuthState.SignedIn(c.session(), c)
            } catch (e: CoreException) {
                AuthState.SignedOut(authErrorMessage(e))
            }
        }
    }
}

File: ios/Pigeon/Auth/AuthViewModel.swift
// AuthViewModel (M5.4) — the thin view-model over the core's session API, the
// iOS mirror of Android's `AuthViewModel.kt`. It owns no protocol/crypto logic:
// it calls the core's async functions, holds the opaque client handle (the token
// stays inside it — Gotcha #1), and publishes UI state.
import SwiftUI
import PigeonCore

/// UI state for the auth flow (mirrors Android's `AuthState`).
enum AuthState {
    /// Checking for a persisted session on launch.
    case restoring
    /// Signed out — showing the form; `error` is the last failure, if any.
    case signedOut(error: String?)
    /// A register/login is in flight.
    case submitting
    /// Signed in. `client` is the opaque core handle the rooms/sync flows hang
    /// off. `signingOut` is true while a logout is in flight; `error` carries a
    /// logout failure that left the session intact (so the user can retry).
    case signedIn(session: Session, client: PigeonClient, signingOut: Bool, error: String?)
}

@MainActor
final class AuthViewModel: ObservableObject {
    @Published private(set) var state: AuthState = .restoring

    // The logged-in client handle; the token stays inside it (in the core). Kept
    // for the flows that hang off it (sync, logout); never unwrapped into
    // app-level secret state.
    private var client: PigeonClient?

    init() {
        Task { await restore() }
    }

    /// Restore a persisted session on launch (offline-first, in the core). A
    /// restore fault must not wedge launch — fall back to the form.
    private func restore() async {
        do {
            if let restored = try await PigeonCore.restoreSession() {
                client = restored
                state = .signedIn(session: restored.session(), client: restored,
                                  signingOut: false, error: nil)
            } else {
                state = .signedOut(error: nil)
            }
        } catch {
            state = .signedOut(error: authErrorMessage(error))
        }
    }

    func login(server: String, username: String, password: String) {
        submit { try await PigeonCore.login(server: server.trimmed, user: username.trimmed, password: password) }
    }

    func register(server: String, username: String, password: String) {
        submit { try await PigeonCore.register(server: server.trimmed, username: username.trimmed, password: password) }
    }

    /// Sign out: revoke the token + clear the persisted session (in the core). On
    /// success drop the handle and return to the form. The core clears local
    /// state even if the server revoke fails; the only error surfaced is a
    /// keystore fault — the session is then still live, so we stay signed in and
    /// show the reason for a retry.
    func logout() {
        guard case let .signedIn(session, _, signingOut, _) = state, !signingOut,
              let c = client else { return }
        state = .signedIn(session: session, client: c, signingOut: true, error: nil)
        Task {
            do {
                try await c.logout()
                client = nil
                state = .signedOut(error: nil)
            } catch {
                state = .signedIn(session: session, client: c, signingOut: false,
                                  error: authErrorMessage(error))
            }
        }
    }

    private func submit(_ call: @escaping () async throws -> PigeonClient) {
        if case .submitting = state { return }
        state = .submitting
        Task {
            do {
                let c = try await call()
                client = c
                state = .signedIn(session: c.session(), client: c, signingOut: false, error: nil)
            } catch {
                state = .signedOut(error: authErrorMessage(error))
            }
        }
    }
}

extension String {
    /// Trim leading/trailing whitespace (server/username fields), matching the
    /// Android VM's `.trim()`.
    var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
}
